Improve overall ore balancing for tendrils_impure.lua

BlueRock2 opened the pull request:
Mining primarily iron and copper veins late game still provides an overabundance of stone and coal.   Would like to change purity from around 82% currently to 86% for iron and copper veins.   Standardized ore weights for each cluster to add up to 1000 so that percentage of each ore = its weight/10.   Coal and stone veins effectively the same percentage since players can tap them or not.   Also kept ore percentages scattered fragments the same since the stone and coal in them is useful early game, but doesn't affect late game much since the scattered fragments are low yield overall and not worth tapping late game.

BlueRock2 committed:
Late game ore balancing for tendrils_impure.lua

Lowered impurity amounts in iron and copper veins as there was little or no need to seek out stone or coal veins later in the game.   Purity went from about 82% to 86%.   Standardized ore weights for each cluster to add up to 1000 so that percentage of each ore = weight / 10.   Coal and stone veins effectively the same since players can tap them or not.   Kept ore percentages scattered fragments the same since the stone and coal in them is useful early game, but doesn't affect late game much since the scattered fragments are low yield overall.

plague006 commented:
Figured I'd drop this here since it's info I have that could potentially be worth knowing. The ratios of ores in vanilla mapgen: https://docs.google.com/spreadsheets/d/1I0PHmQyMlNjZ7t4tdJ_Mdgl9oM1ao6yG3FAY10goRts/edit#gid=1176048894

BlueRock2 commented:
On my first submission for tendrils_impure ore generation, I did some testing by clearing void on a 500,000 square tile area several times and adding up the ore counts.   But it isn't very accurate because I was manually hovering over the tendrils and adding up approximate amounts.    Doing this however, tends not to reveal the amounts added by impurities in the veins.   I think I may know enough about LUA now to have it accurately add up all the ore of each type within a square area for me, but would like to get my hands on existing code if it is already available somewhere.

plague006 commented:
In your script-output folder you'll get `resource_dump.lua` containing a table with all resources.
```lua
game.write_file('resource_dump.lua', serpent.block(game.player.surface.get_resource_counts()))
```

BlueRock2 commented:
Thanks!, I will do some resource counts on several map generations tonight, as I believe you are right that the Diggy total tendril ore percentages need to be close to default map generation.  Except for maybe stone, which because of constructing all the supports would need to be slightly higher.   May still have to go by feel on that one.

plague006 commented:
>  May still have to go by feel

I think that's true overall. I'm not offering vanilla numbers to say they should be more or less like them. In vanilla the vast majority of coal and stone patches go unused. It's just for comparison.

If you want a ratio to potentially aim for, I would recommend checking consumed items consumed in a completed diggy save.

BlueRock2 commented:
Taking into account past consumption sounds like a really good idea.   The Factorio default ore percentages appear to be based on what you would use if you kept cranking out equal amounts of all science packs over time.  But currently most players spend the most time in game before then.   I will gather the consumption numbers, but I think they may have changed noticeably each time depending on alien difficulty (more iron for bullets) or how hard it is to get uranium (delays switch from coal to nuclear power).   I think the optimal target resources is probably somewhere between current average consumption per Diggy game and that of resources needed for infinite science.

BlueRock2 commented:
Just for reference while I do further testing to verify ore averages for this PR:   Diggy consumption averages for the currently playing game (close to end game) and last three end game saves normalized with iron =1 is copper = 0.59,  coal = 0.41 and stone = 0.20.    Interestingly I found that the default ore values you gave me are right on the nose for iron/copper ratio for infinite science (1 to 0.812).   But you are right, coal and stone in the default ratios appear much higher than needed and means a lot goes unused.   They must have did that to give people flexibility early/mid game, because infinite science production takes a lot lower ratio.   For example coal needed for infinite science is only 1/10 the amount of iron (if nuclear and electric furnaces are used)

BlueRock2 commented:
After some testing, overall map ore averages for this PR come out to iron = 1,  copper = 0.75,  coal = 0.38 and stone = .30.    I think the iron/copper/coal ratio is just about right now, but taking stone down just a little more and re-testing.

BlueRock2 committed:
Further balance tendrils_impure.lua ore ratios

This update is to better match overall map ore ratios in tendrils_impure.lua to ore consumption ratios on previous Diggy games.   Fragmented tendril ratios target anticipated early game needs, while the larger tendrils help balance out expected mid/late game ratios.   Toatal map ore average ratios for this commit are iron/copper/coal/stone equal to approx 1/0.8/0.4/0.25

BlueRock2 commented:
Made slight tweeks, mostly concerning stone this last revision.   Ran 10 map generations on 707x707 area (approx half million tiles) and captured ore counts.   Map ore generation ratio averages came out to iron/copper/coal/stone as 1/0.80/0.40/0.25.    I believe these are good numbers.

plague006 commented:
Looks good to me, merging with my thanks.
